### PR TITLE
Enable `testIncompleteListenerDecl` test

### LINKED ISSUE #27778: Enable testIncompleteListenerDecl
**Description:**
Disabled the above test temporarily to fix it

### LINKED ISSUE #28246: Parser crashes for let expression as default value in an object type descriptor
**Description:**
$title. Fields in an object type descriptor cannot have a default value.

**Steps to reproduce:**
The following produces compilation errors.
```ballerina
int x = 1;

type Person object {
    public int id = x + 10;
};
```

```cmd
ERROR [qwe.bal:(4:19,4:19)] missing at token
ERROR [qwe.bal:(4:19,4:19)] missing semicolon token
ERROR [qwe.bal:(4:19,4:19)] undefined annotation '$missingNode$_1'
ERROR [qwe.bal:(4:21,4:22)] invalid token '='
ERROR [qwe.bal:(4:21,4:22)] unknown type 'x'
ERROR [qwe.bal:(4:23,4:23)] missing identifier
ERROR [qwe.bal:(4:23,4:23)] missing semicolon token
ERROR [qwe.bal:(4:27,4:27)] missing identifier
```

But when a let expression is used in the default value, the parser crashes.
```ballerina
type Person object {
    public int id = let int x = 3 in 2 * x;
};
```

```log
[2021-01-27 11:47:47,554] SEVERE {b7a.log.crash} - defaultable-parameter 
java.lang.IllegalStateException: defaultable-parameter
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.getNextRuleForSemicolon(BallerinaParserErrorHandler.java:4396)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.getNextRule(BallerinaParserErrorHandler.java:2552)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1249)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:380)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:328)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1246)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:380)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.fixAndContinue(AbstractParserErrorHandler.java:328)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1246)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1872)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1235)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekInAlternativesPaths(AbstractParserErrorHandler.java:248)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatchInAlternativePaths(BallerinaParserErrorHandler.java:1872)
	at io.ballerina.compiler.internal.parser.BallerinaParserErrorHandler.seekMatch(BallerinaParserErrorHandler.java:1235)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatchInSubTree(AbstractParserErrorHandler.java:199)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.seekMatch(AbstractParserErrorHandler.java:181)
	at io.ballerina.compiler.internal.parser.AbstractParserErrorHandler.recover(AbstractParserErrorHandler.java:102)
	at io.ballerina.compiler.internal.parser.AbstractParser.recover(AbstractParser.java:106)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseParameter(BallerinaParser.java:2017)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseParamList(BallerinaParser.java:1864)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseFuncSignature(BallerinaParser.java:1489)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseFunctionKeywordRhs(BallerinaParser.java:1117)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseFunctionKeywordRhs(BallerinaParser.java:1124)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseFuncDefOrFuncTypeDesc(BallerinaParser.java:1099)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseObjectMethodOrFuncTypeDesc(BallerinaParser.java:6130)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseObjectMemberWithoutMeta(BallerinaParser.java:5971)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseObjectMemberWithoutMeta(BallerinaParser.java:5946)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseObjectMember(BallerinaParser.java:5932)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseObjectMember(BallerinaParser.java:5929)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseObjectMember(BallerinaParser.java:5929)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseObjectMembers(BallerinaParser.java:5859)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseObjectTypeDescriptor(BallerinaParser.java:5769)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTypeDescriptorInternal(BallerinaParser.java:2348)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTypeDescriptorInternal(BallerinaParser.java:2237)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTypeDescriptor(BallerinaParser.java:2230)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTypeDescriptor(BallerinaParser.java:2224)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTypeDescriptor(BallerinaParser.java:2210)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseModuleTypeDefinition(BallerinaParser.java:3151)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:778)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:760)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:316)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseTopLevelNode(BallerinaParser.java:249)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parseCompUnit(BallerinaParser.java:153)
	at io.ballerina.compiler.internal.parser.BallerinaParser.parse(BallerinaParser.java:98)
	at io.ballerina.compiler.syntax.tree.SyntaxTree.from(SyntaxTree.java:55)
	at io.ballerina.projects.DocumentContext.parse(DocumentContext.java:86)
	at io.ballerina.projects.DocumentContext.syntaxTree(DocumentContext.java:90)
	at io.ballerina.projects.DocumentContext.getModuleLoadRequests(DocumentContext.java:127)
	at io.ballerina.projects.DocumentContext.moduleLoadRequests(DocumentContext.java:121)
	at io.ballerina.projects.ModuleContext.populateModuleLoadRequests(ModuleContext.java:171)
	at io.ballerina.projects.PackageResolution.getModuleLoadRequestsOfDirectDependencies(PackageResolution.java:160)
	at io.ballerina.projects.PackageResolution.createDependencyGraphFromSources(PackageResolution.java:229)
	at io.ballerina.projects.PackageResolution.buildDependencyGraph(PackageResolution.java:146)
	at io.ballerina.projects.PackageResolution.<init>(PackageResolution.java:70)
	at io.ballerina.projects.PackageResolution.from(PackageResolution.java:77)
	at io.ballerina.projects.PackageContext.getResolution(PackageContext.java:201)
	at io.ballerina.projects.PackageCompilation.from(PackageCompilation.java:93)
	at io.ballerina.projects.PackageContext.getPackageCompilation(PackageContext.java:194)
	at io.ballerina.projects.Package.getCompilation(Package.java:124)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:68)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:155)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
```

**Affected Versions:**
slalpha1

### PR DESCRIPTION
**Pull request by chiranSachintha**
Fixes #27778 
Fixes #28246 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -2008,9 +2008,13 @@ private STNode parseParameter(SyntaxKind prevParamKind, boolean isParamNameOptio
                 }
 
                 STToken token = peek();
-                Solution solution =
-                        recover(token, ParserRuleContext.PARAMETER_START, prevParamKind, isParamNameOptional);
-
+                Solution solution;
+                if (prevParamKind == SyntaxKind.DEFAULTABLE_PARAM) {
+                    solution = recover(token, ParserRuleContext.PARAMETER_START_WITHOUT_ASTERISK,
+                                       prevParamKind, isParamNameOptional);
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:2014]: According to spec, a `defaultable-param` can be followed by an `included-record-param` right. So, do we need to skip asterisk here?

Comment (chiranSachintha) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java:2014]: This should be allowed. Create a separate issue for this.(https://github.com/ballerina-platform/ballerina-lang/issues/27818)

Comment (chiranSachintha) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java]: Fixed.

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 2]
```diff
@@ -3987,6 +3987,7 @@ private boolean isInTypeDescContext() {
             case TYPE_DESC_IN_SERVICE:
             case TYPE_DESC_IN_PATH_PARAM:
             case STMT_START_BRACKETED_LIST:
+            case DEFAULTABLE_PARAM:
```
Comment (lochana-chathura) [inline compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java:3990]: DEFAULTABLE_PARAM is not a part of the type-descriptor. Defining it like this is wrong.

**Review by lochana-chathura — CHANGES_REQUESTED** [review]
I had a look on this issue. Problem is we reach getnextrule for `ASTERISK` not only under `REQUIRED_PARAM` ctx but also under `DEFAULTABLE_PARAM` ctx. 

i.e. Need to give next rule as `TYPE_DESC_IN_PARAM` for `DEFAULTABLE_PARAM` as well.
https://github.com/ballerina-platform/ballerina-lang/blob/b67f907b5701940c33b726a2f6fb575cc191019a/compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java#L2616-L2617

File: compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java (hunk 2 above)
Comment (chiranSachintha) [reply compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParserErrorHandler.java]: Fixed.

**Review by lochana-chathura — DISMISSED** [review]

**Comment by MaryamZi** [comment]
Shall we add a test for the scenario in https://github.com/ballerina-platform/ballerina-lang/issues/28246 too?

**Comment by lochana-chathura** [comment]
Would be good to add one of the above test cases as a parser test as well. Since the fixed issue is from the parser.

**Comment by chiranSachintha** [comment]
> Would be good to add one of the above test cases as a parser test as well. Since the fixed issue is from the parser.

Added.

**Review by MaryamZi — APPROVED** [review]

**Review by lochana-chathura — APPROVED** [review]

**Pull request merged**